test, upcoming: [M3-7134, M3-7314] - Add integration test for AGLB Edit Certificate flow and improve form validation

mjac0bs opened the pull request:
## Description 📝
This PR does two things, because fixing the validation was necessary for writing a complete integration test that confirmed all input errors were handled gracefully.

1. Add Cypress integration tests to confirm the UI flow for AGLB certification editing works as expected.
2. Provide form validation to prevent a user from being able to edit a certificate and provide a blank label.

## Changes  🔄
Integration tests in `load-balancer-certificates.spec.ts` test the following:
- User can edit certificates from the AGLB certificate page (route: /loadbalancers/:id/certificates)
- Certificates page updates to reflect edited certificates
- Cloud Manager handles user input errors (e.g. empty or invalid fields/selections) gracefully
- Cloud Manager handles server errors (e.g. HTTP 500 responses from the API) gracefully

In the Cloud Manager UI:
- A validation error ("Label must not be empty.") is provided when a user attempts to submit the Edit Certificate drawer with a blank label field. 

## Preview 📷
| Before  | After   |
| ------- | ------- |
| <video src="https://github.com/linode/manager/assets/114685994/aabf207c-6bf0-451c-8d0a-c580143f7ba1" /> | <video src="https://github.com/linode/manager/assets/114685994/94143886-e6f4-459e-bc30-d6bc79ce10c6" /> |

![Screenshot 2023-11-06 at 2 39 18 PM](https://github.com/linode/manager/assets/114685994/25c1b7a7-3f93-4d2c-ac20-0da6d3a28ea1)

## How to test 🧪

### Prerequisites
(How to setup test environment)
- Checkout this PR.
- Make sure the `aglb` feature flag is on and **turn mocks on**. 

### Reproduction steps
(How to reproduce the issue, if applicable)
- Checkout `develop`.
- Go to http://localhost:3000/loadbalancers/0/certificates.
- Use the action menu (`...`) to select the Edit action and open the Edit Certificate drawer.
- Clear the pre-populated `label` field.
- Submit the form.
- Observe there is no validation error and in the request payload, the blank label is submitted. (Due to the mocks, this isn't visible in the UI.)

### Verification steps 
(How to verify changes)
- On this branch, go to http://localhost:3000/loadbalancers/0/certificates.
- Use the action menu (`...`) to select the Edit action and open the Edit Certificate drawer.
- Clear the pre-populated `label` field.
- Submit the form.
- Observe there is a validation error `Label must not be empty` and the form will not submit until a label is provided.

Also confirm the following validation still works as expected by checking the request payload:
- When editing a TLS certificate:
   - A user can edit just the form label, and the form will submit. The certificate will be excluded from the payload since it has not changed.
   - A user can edit just the key, and the form will submit. The existing certificate will be included in the payload, since updating a key requires a cert.
   - A user _cannot_ edit just the cert and will see a validation error on Private Key. This is invalid because updating a cert requires a key, and API does not return the key for security. 
   - A user can edit both the cert and the key, and the form will submit. Both the edited cert and edited key will be in the payload.

Confirm the integration test passes:
```
yarn cy:run -s "cypress/e2e/core/loadBalancers/load-balancer-certificates.spec.ts"
```
## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

mjac0bs commented on packages/manager/src/features/LoadBalancers/LoadBalancerDetail/Certificates/EditCertificateDrawer.tsx:
```diff
@@ -89,25 +91,24 @@ export const EditCertificateDrawer = (props: Props) => {
         <Notice variant="error">Error loading certificate.</Notice>
       ) : (
         <form onSubmit={formik.handleSubmit}>
-          {errorMap.none && <Notice text={errorMap.none} variant="error" />}
```
Somehow an extra error notice had slipped in originally -- cleaned this up here.

mjac0bs commented on packages/manager/src/features/LoadBalancers/LoadBalancerDetail/Certificates/EditCertificateDrawer.tsx:
```diff
@@ -55,26 +55,28 @@ export const EditCertificateDrawer = (props: Props) => {
     },
+    // Disabling validateOnBlur and validateOnChange when an API error is shown prevents
+    // all API errors from disappearing when one field is changed.
+    validateOnBlur: !error,
+    validateOnChange: !error,
```
Following suit with what we did in another edit drawer even though this doesn't look as nice, it seems better than have disappearing errors if a user were to clear the label and update the cert, but not provide a key (see PR description video).

bnussman-akamai reviewed (approved):
Changes look great! Thanks for all of the tests too 🧪

coliu-akamai reviewed (commented):
Looks good! Confirmed everything listed, just had one small comment - 

Could we add a scroll to error, especially for the forms that have the Private Key field?

<video src="https://github.com/linode/manager/assets/139280159/10c372a8-0a4d-4695-bdb7-57b24a003954" />
-----
Confirmed that all tests pass! ✅ 
Confirmed that error appears! ✅ 
Confirmed previous validation: ✅ 
- ✅  A user can edit just the form label, and the form will submit. The certificate will be excluded from the payload since it has not changed.
- ✅  A user can edit just the key, and the form will submit. The existing certificate will be included in the payload, since updating a key requires a cert.
- ✅ A user cannot edit just the cert and will see a validation error on Private Key. This is invalid because updating a cert requires a key, and API does not return the key for security.
- ✅ A user can edit both the cert and the key, and the form will submit. Both the edited cert and edited key will be in the payload.

dwiley-akamai reviewed (approved):
Integration test passes ✅

"Label must not be empty" validation error when attempting to submit an edited cert with no label ✅

Edit just the form label, and the form will submit. The certificate will be excluded from the payload since it has not changed. ✅

Edit just the key, and the form will submit. The existing certificate will be included in the payload, since updating a key requires a cert. ✅

Cannot edit just the cert. See a validation error on Private Key. ✅

Can edit both the cert and the key, and the form will submit. Both the edited cert and edited key are in the payload. ✅

mjac0bs commented:
> Could we add a scroll to error, especially for the forms that have the Private Key field?

@coliu-akamai Good call out for smaller screens!

I implemented scroll-into-view for the Edit drawer in [423e64c](https://github.com/linode/manager/pull/9880/commits/423e64c2b2f7e8e273f02f98f5f80ff68ef6115d). 

https://github.com/linode/manager/assets/114685994/966fe97c-3c8e-4088-8e8c-ade7a5363e29

Since we'd want the same behavior in the Create drawer, so I made the change there too. (Also noticed that the Create drawer is not an expanded width, which is a bad UX when the user pastes long certs, so I adjusted in [7624cec](https://github.com/linode/manager/pull/9880/commits/7624cec42575d18350b691b34cd6b7222f80516c).)

https://github.com/linode/manager/assets/114685994/47370313-d08f-4da8-aacf-37adb5ca001c

coliu-akamai reviewed (approved):
awesome, thanks Mariah! 🎉
